File: java/shared/java/org/mochi/libmochi/zkgroup/groups/GroupPublicParams.java
//
// Copyright 2020-2021 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.mochi.libmochi.zkgroup.groups;

import static org.mochi.libmochi.internal.FilterExceptions.filterExceptions;

import org.mochi.libmochi.internal.Native;
import org.mochi.libmochi.zkgroup.InvalidInputException;
import org.mochi.libmochi.zkgroup.internal.ByteArray;

public final class GroupPublicParams extends ByteArray {

  public GroupPublicParams(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class, () -> Native.GroupPublicParams_CheckValidContents(contents));
  }

  public GroupIdentifier getGroupIdentifier() {
    byte[] newContents = Native.GroupPublicParams_GetGroupIdentifier(contents);

    try {
      return new GroupIdentifier(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}

File: swift/Sources/LibMochiClient/zkgroup/GroupPublicParams.swift
//
// Copyright 2020-2021 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import MochiFfi

public class GroupPublicParams: ByteArray {
    public required init(contents: [UInt8]) throws {
        try super.init(contents, checkValid: mochi_group_public_params_check_valid_contents)
    }

    public func getGroupIdentifier() throws -> GroupIdentifier {
        return try withUnsafePointerToSerialized { contents in
            try invokeFnReturningSerialized {
                mochi_group_public_params_get_group_identifier($0, contents)
            }
        }
    }
}
